### ISSUE TITLE
Need more documentation about pullSecrets

### ISSUE THREAD

**Issue by pwillis-eiq** [devspace-sh/devspace#2504]
**Is your feature request related to a problem?**

Yes. I am trying to build an image in a private registry (GCR). I have configured the devspace.yaml:

```
images:
  python:
    image: gcr.io/MY-PROJECT/devspace-python:latest
    dockerfile: ./containers/python/Dockerfile

deployments:
  app:
    helm:
      chart:
        name: component-chart
        repo: https://charts.devspace.sh
      values:
        containers:
          - image: username/app
```

And I run `devspace deploy`:

```
$ devspace deploy
info Using namespace 'ds-devspace'
info Using kube context 'CONTEXT'
Ensuring image pull secret for registry: gcr.io...
Couldn't retrieve username for registry gcr.io from docker store
Couldn't retrieve password for registry gcr.io from docker store
Couldn't find a running docker daemon. Will fallback to kaniko
build:python Rebuild image gcr.io/MY-PROJECT/devspace-python because tag is missing
build:python Building image 'gcr.io/MY-PROJECT/devspace-python:e11fe71' with engine 'kaniko'
build:python Waiting for build init container to start...
```

And it hangs forever, because the initContainer won't finish:

```
$ kubectl get pods
NAME                            READY   STATUS     RESTARTS   AGE
app-devspace-7ffb887865-9j4hx   1/1     Running    0          117m
devspace-build-kaniko-zsn7k     0/1     Init:0/1   0          59s
$ kubectl describe pod devspace-build-kaniko-zsn7k
Name:         devspace-build-kaniko-zsn7k
Namespace:    ds-devspace
Priority:     0
Node:         MY-NODE/10.0.0.27
Start Time:   Sun, 22 Jan 2023 13:00:07 -0500
Labels:       devspace-build=true
              devspace-build-id=tbgjxumpcqqq
              devspace-pid=oehvwbyfbuom
Annotations:  <none>
Status:       Pending
IP:
IPs:          <none>
Init Containers:
  context:
    Container ID:
    Image:         alpine
    Image ID:
    Port:          <none>
    Host Port:     <none>
    Command:
      sh
    Args:
      -c
      while [ ! -f /tmp/done ]; do sleep 2; done
    State:          Waiting
      Reason:       PodInitializing
    Ready:          False
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /context from context (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-5c86t (ro)
Containers:
  kaniko:
    Container ID:
    Image:         gcr.io/kaniko-project/executor:v1.8.1
    Image ID:
    Port:          <none>
    Host Port:     <none>
    Args:
      --dockerfile=/context/Dockerfile
      --context=dir:///context
      --destination=gcr.io/MY-PROJECT/devspace-python:e11fe71
      --snapshotMode=time
      --cache=true
      --cache-repo=gcr.io/MY-PROJECT/devspace-python
    State:          Waiting
      Reason:       PodInitializing
    Ready:          False
    Restart Count:  0
    Environment:
      DD_AGENT_HOST:   (v1:status.hostIP)
      DD_ENTITY_ID:    (v1:metadata.uid)
    Mounts:
      /context from context (rw)
      /kaniko/.docker from devspace-pull-secrets (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-5c86t (ro)
Conditions:
  Type              Status
  Initialized       False
  Ready             False
  ContainersReady   False
  PodScheduled      True
Volumes:
  context:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
    SizeLimit:  <unset>
  devspace-pull-secrets:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  devspace-pull-secrets
    Optional:    false
  kube-api-access-5c86t:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason       Age                From               Message
  ----     ------       ----               ----               -------
  Normal   Scheduled    94s                default-scheduler  Successfully assigned ds-devspace/devspace-build-kaniko-zsn7k to MY-NODE
  Warning  FailedMount  30s (x8 over 93s)  kubelet            MountVolume.SetUp failed for volume "devspace-pull-secrets" : secret "devspace-pull-secrets" not found
```

So it's because there's no devspace pull secrets, okay...

So I go look up the docs for pull secrets: [https://www.devspace.sh/docs/configuration/pullSecrets/](https://www.devspace.sh/docs/configuration/pullSecrets/)

And I see I can specify a registry, username and password. But I'm not sure how to make that work with Google Cloud Container Registry (or, well, any access method that isn't a Docker registry). Luckily Google Cloud has docs on how to generate Docker credentials, but ideally I would be able to read the DevSpace docs and figure out how to solve the problem above myself.

I see in the docs I can set more options:

```
email string 
The optional email to use
```

What is this email for? When would I use it?

```
secret string 
The secret to create
```

What secret to create? How is this used? When would I use this? Would I need to do anything else, reference something else, etc?

```
serviceAccounts string[] 
The service account to add the secret to
```

What service account? Like a kubernetes service account? When would I use this, and how?

**Which solution do you suggest?**

Please provide more documentation about the configuration options for pullSecrets so I know when (or if) to use these settings, and how.

Thank you

**Which alternative solutions exist?**

**Additional context**

I want my development environment to depend on local tools as little as possible (notably: not having to install Docker Desktop). So I would *love* to use Kaniko to build images. Whatever I need to do to use Kaniko for image building, and using Google Cloud Container Registry for the images, would be great. Thanks

**Comment by lizardruss** [comment]
Hello! Sorry to hear that you're having difficulty. The newer version of DevSpace no longer falls back to Kaniko, and instead uses docker or buildkit for image  building and the [localRegistry](https://www.devspace.sh/docs/configuration/localRegistry/) feature to make images available to the cluster. This doesn't help your case, since you'd like to continue using Kaniko.

For Kaniko, DevSpace relies on pulling the registry credentials from the local docker credential store to correctly configure the service accounts for the Kaniko build pods. For GCR they strongly recommend using [gcloud credential helper](https://cloud.google.com/container-registry/docs/advanced-authentication#gcloud-helper), which configures the docker credential store. Opting out of this means you'll need to provide the `username` and `password`, according to the cloud providers instructions. I believe for GCP this would involve using an [access token](https://cloud.google.com/container-registry/docs/advanced-authentication#token), however this is a short lived token, so you may need to continually refresh it.

Regarding your questions about the different configuration options, they become a little more clear when you understand the options for [creating a Kubernetes docker registry secret.](https://kubernetes.io/docs/concepts/containers/images/#creating-a-secret-with-a-docker-config) The `email` option corresponds to the `--docker-email` flag, and `secret` maps to the `<name>` of the secret. You probably don't need to set these, since the defaults should be sufficient.

Likewise `serviceAccounts string[]` refers to [adding the registry pull secret to one or more service accounts](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#add-imagepullsecrets-to-a-service-account). The `serviceAccounts` option is only needed if you want to use a different service account. DevSpace will automatically use `default` unless you specify otherwise. As for when you would use this and how, that is probably best covered by the kubernetes docs [Configure Service Accounts for Pods](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/).

Hopefully this adds some clarity. I'll discuss with my colleagues about what documentation updates we'd like to make.

**Comment by peterwwillis** [comment]
@lizardruss there appears to still be a lack of documentation somewhere, as I've run into this problem again without an apparent solution. I assume other users would find themselves in the same situation

 - I have a GKE cluster in GCloud
 - I select a new empty namespace
 - I create a devspace config, and configure it to use Kaniko for image building, but I do not need it to use a registry because I just want to rebuild public images.
 - I run `devspace build` and wait for the image to build
 - The build hangs indefinitely because Devspace refuses to create a secret `devspace-pull-secrets` which it is trying to mount, so the pod is stuck in Init mode.

The thing is... I have no need for build secrets. I want to pull the Docker Hub public ubuntu image, apply some custom Dockerfile lines, and then run the new image. So I don't need to access any Docker registry, so secrets are not needed. But Devspace is hanging anyway.

[Here](https://www.devspace.sh/docs/configuration/pullSecrets/) it says:

> For [images](https://www.devspace.sh/docs/configuration/images/), DevSpace will also automatically create a pull secret, if authentication data can be found in the local docker store and creation is [enabled](https://www.devspace.sh/docs/configuration/images/#8-create-pull-secrets) via createPullSecret.

However, even after specifying `createPullSecret: false` in my image section, it still tries to use the missing secret and hangs forever.

Sample devspace config:

```
version: v2beta1
name: mydevspace

pipelines:
  dev:
    run: |-
      GIT_TAG_OR_HASH=$(git describe --always)
      run_dependencies --all
      #ensure_pull_secrets --all # <-- commented out
      build_images --all -t $GIT_TAG_OR_HASH
      create_deployments --all
      start_dev app

images:
  app:
    image: ubuntu:23.10
    dockerfile: ./Dockerfile
    context: ./
    kaniko:
      cache: true
    appendDockerfileInstructions: ["ADD install_tooling.sh .", "RUN chmod +x install_tooling.sh && ./install_tooling.sh && rm install_tooling.sh"]
    rebuildStrategy: ignoreContextChanges
    skipPush: true
    createPullSecret: false

deployments:
  app:
    helm:
      chart:
        name: component-chart
        repo: https://charts.devspace.sh
      values:
        containers:
          - image: ubuntu:23.10

dev:
  app:
    # Search for the container that runs this image
    imageSelector: ubuntu:23.10
    sync:
      - path: ./:/home/ubuntu/devspace
    terminal:
      command: /home/ubuntu/devspace/devspace_start.sh
    ssh:
      enabled: true
    proxyCommands:
      - command: devspace
      #- command: kubectl
      #- command: helm
      - gitCredentials: true
```



I think there are 3 bugs here:

1. If I don't need a secret to complete these operations, Devspace shouldn't refuse to run without one
2. Even if it does require it, it should die and tell me that, and not hang indefinitely
3. If I follow the docs and try to disable the pull secret, it should not try to use the missing/inoperable pull secret anyway

**Comment by peterwwillis** [comment]
Following up, I found `skipPullSecretMount` in the kaniko configuration block; setting it to true led to the error
```
build_images: build images: error building image ubuntu:da47967: waiting for kaniko: kaniko pod pxw0bhy/devspace-build-kaniko-svwn8 has unexpectedly exited with code 1: error checking push permissions -- make sure you entered the correct tag name, and that you are authenticated correctly, and try again: checking push permission for "ubuntu:da47967": POST https://index.docker.io/v2/library/ubuntu/blobs/uploads/: UNAUTHORIZED: authentication required; [map[Action:pull Class: Name:library/ubuntu Type:repository] map[Action:push Class: Name:library/ubuntu Type:repository]]
```

Reading the documentation further, it seems that I can only disable pushing if I don't use kaniko, which is disappointing:

> SkipPush will not push the image to a registry if enabled. Only works if docker or buildkit is chosen as build method

Our purpose in using Kaniko (and DevSpace in general) is to avoid having to install and maintain a Docker Desktop install on laptops, and instead use Kubernetes for development. So I would request that skipPush be supported on kaniko as well, or a workaround be documented. (I'm assuming I'll have to set up some kind of self-hosted Docker registry as a Helm deployment, deploy that into the cluster, and point Kaniko at that, just so it can push the image back where it came from, so that I can get the build to work.......)

**Comment by peterwwillis** [comment]
@lizardruss that does seem to work, thank you!

---

Fwiw, there does seem to be a bunch of behavior that doesn't seem to exist in the documentation, that I'm learning by experimenting.

I went back to the original config I wanted to use, and that does seem to work by default now that it's not trying to use kaniko. (I think I've been confused that kaniko was not the builder for localRegistry, but buildkitd as you say; was not familiar with that container/system as a remote/k8s builder).

I think it's worth clarifying that if the user relies on the localRegistry functionality, they shouldn't specify any other build engine in the image configuration (such as kaniko) or it'll prevent the `localregistry` engine from building. I thought I could add:
```
    kaniko: {}
```
and that it would just revert back to localRegistry support, but that doesn't happen; leaving out the build engine makes it work with localRegistry.

---

If the local Docker CLI tool has a connection to a remote Docker daemon, Devspace will attempt to generate Docker pull secrets in K8s and fail. However, if I disable the connection to the Docker daemon (I use colima for the Docker daemon), it does generate a docker pull secret in K8s successfully.


---

If I try to just run devspace on a public image, using the localRegistry support, it will look for a Dockerfile and fail if it doesn't exist:

```
version: v2beta1

images:
  python:
    image: ubuntu:23.04

dev:
  my-dev:
    imageSelector: ubuntu:23.04
    terminal: {}
```

```
[user@hostname devspace (main)]$ devspace dev
info Using namespace 'user'
info Using kube context 'context'
Ensuring image pull secret for registry: hub.docker.com...
Couldn't retrieve username for registry  from docker store
Couldn't retrieve password for registry  from docker store
local-registry: Starting Local Image Registry
build_images: build images: error during shouldRebuild check: Dockerfile /Users/user/devspace/Dockerfile missing: stat /Users/user/devspace/Dockerfile: no such file or directory
fatal exit status 1
```

Based on the docs (https://www.devspace.sh/docs/configuration/images/build#build-configuration) it seems that `dockerfile` is optional, but if a Dockerfile doesn't exist, `devspace dev` fails. I don't see a dockerfile option required in the `devspace dev` doc either (https://www.devspace.sh/docs/configuration/dev/#config-reference). I guess the *option* is optional, but the *file* is not optional?

If I add a Dockerfile, it will seemingly ignore the ubuntu image, then use whatever is in the Dockerfile. However, it will re-use the last build rather than use a new build if the Dockerfile changes, so subsequent commands rebuild the image, but don't result in a new dev container when the Dockerfile changes.

To deal with that, I tried explicitly passing the tag in a pipelines section:

```
version: v2beta1

pipelines:
  dev: |-
    run_dependencies --all
    ensure_pull_secrets --all
    build_images --all -t $(git describe --always)
    create_deployments --all
    start_dev --all

images:
  python:
    image: ubuntu:23.04

dev:
  my-dev:
    imageSelector: ubuntu:23.04
    terminal: {}
```

And that leads to another hanging situation, even though the secrets it's looking for actually exist:

```
01:24:46 dev:my-dev Update replaced deployment with patch:
 {"metadata":{"annotations":{"deployment.kubernetes.io/revision":"2","devspace.sh/container":null}},"spec":{"template":{"metadata":{"annotations":{"devspace.sh/container":null,"devspace.sh/imageSelector":"localhost:31886/ubuntu:9abb179"}}}}}
01:24:46 dev:my-dev Waiting for pod to become ready...
01:24:46 dev:my-dev Start selecting a single container with selector image selector: localhost:31886/ubuntu:9abb179
01:24:57 dev:my-dev Pod app-59fbcd5cff-ljfzr: Back-off restarting failed container (BackOff)
01:24:57 dev:my-dev Pod devspace-build-kaniko-kfcjz: Unable to attach or mount volumes: unmounted volumes=[devspace-pull-secrets], unattached volumes=[kube-api-access-2snkh devspace-pull-secrets context]: timed out waiting for the condition (FailedMount)
01:24:57 dev:my-dev Pod devspace-build-kaniko-kfcjz: Unable to attach or mount volumes: unmounted volumes=[devspace-pull-secrets], unattached volumes=[devspace-pull-secrets context kube-api-access-2snkh]: timed out waiting for the condition (FailedMount)
01:24:57 dev:my-dev Pod devspace-build-kaniko-kfcjz: MountVolume.SetUp failed for volume "devspace-pull-secrets" : secret "devspace-pull-secrets" not found (FailedMount)
01:24:57 dev:my-dev DevSpace still couldn't find any Pods that match the selector. DevSpace will continue waiting, but this operation might timeout
01:25:07 dev:my-dev DevSpace still couldn't find any Pods that match the selector. DevSpace will continue waiting, but this operation might timeout
01:25:18 dev:my-dev DevSpace still couldn't find any Pods that match the selector. DevSpace will continue waiting, but this operation might timeout
01:25:28 dev:my-dev DevSpace still couldn't find any Pods that match the selector. DevSpace will continue waiting, but this operation might timeout
01:26:10 dev:my-dev DevSpace still couldn't find any Pods that match the selector. DevSpace 
```

Turns out that was due to a pre-existing deployment (devspace operations aren't immutable or idempotent). Did a purge, tried again, complained about no deployment. So I added a deployment section and tried again, and that worked, at first.

But changing the Dockerfile again still results in it re-using the last cached build. Later it turns out that if I just leave the tag off the `ubuntu:23.04` on all the lines in the config, it works as expected, deploying the currently-built image after the Dockerfile changes. No pipeline or build tag needed, but a deployment section is.

---

As I'm looking for documentation, I'll use your website's search box in the top right hand corner. If I put in "labelSelector", and click the first result, it brings me to the documentation for version 5.x. But I don't realize that until much later, when I realize I can't find this page from the main docs page. The next link after the first one in the search box is for 4.x. No indication from the search box that it's linking to older docs. (incidentally, the 5.x docs are actually easier to read and understand than the 6.x docs)

labelSelector doesn't even seem to be documented for 6.x: https://www.devspace.sh/docs/configuration/dev/selectors/image#dev-labelSelector



---

It does seem like there's quite a learning curve to understand how Devspace wants the user to use it, guessing what Devspace doesn't like, trying another thing...

I really do like this tool, and think it could be a game changer for a lot of teams. But it's taking us weeks/months to work through all the idiosyncrasies to the point where we actually know how to use this tool and can introduce it to our other teams. I've been using it for at least a month straight and I still can't get it to do what I want (just drop me into an Ubuntu container so I can do cloud development work without Docker Desktop installed) without weeks of experimentation and research.

(for posterity, if anyone is looking for the solution, the following config appears to be required; if any lines are missing or different, you may end up with old cached images, or something else won't work:)

```
version: v2beta1

images:
  ubuntu:
    image: nonsense
    rebuildStrategy: always

deployments:
  app:
    helm:
      chart:
        name: component-chart
        repo: https://charts.devspace.sh
      values:
        containers:
          - image: nonsense

dev:
  my-dev:
    imageSelector: nonsense
    terminal: {}
```

**Comment by lizardruss** [comment]
> 
> @lizardruss that does seem to work, thank you!

Glad this got you unblocked!

> 
> Fwiw, there does seem to be a bunch of behavior that doesn't seem to exist in the documentation, that I'm learning by experimenting.
> 
> I went back to the original config I wanted to use, and that does seem to work by default now that it's not trying to use kaniko. (I think I've been confused that kaniko was not the builder for localRegistry, but buildkitd as you say; was not familiar with that container/system as a remote/k8s builder).
> 
> I think it's worth clarifying that if the user relies on the localRegistry functionality, they shouldn't specify any other build engine in the image configuration (such as kaniko) or it'll prevent the `localregistry` engine from building. I thought I could add:
> 
> ```
>     kaniko: {}
> ```
> 
> and that it would just revert back to localRegistry support, but that doesn't happen; leaving out the build engine makes it work with localRegistry.


I agree that this is confusing. The localRegistry feature will use buildkit as the build engine for both local or remote image builds to help simplify configuration. It was previously implemented to use the build engine specified under `images`, but it was only intended to work with `docker` & `buildkit` engines. In the future it might be nice to get it working with different build engines, but it's sometimes not compatible (`kaniko` is a good example, where deep knowledge of kaniko configuration is needed to get it working correctly with the local registry)

> If the local Docker CLI tool has a connection to a remote Docker daemon, Devspace will attempt to generate Docker pull secrets in K8s and fail. However, if I disable the connection to the Docker daemon (I use colima for the Docker daemon), it does generate a docker pull secret in K8s successfully.
> 
> If I try to just run devspace on a public image, using the localRegistry support, it will look for a Dockerfile and fail if it doesn't exist:
> 
> ```
> version: v2beta1
> 
> images:
>   python:
>     image: ubuntu:23.04
> 
> dev:
>   my-dev:
>     imageSelector: ubuntu:23.04
>     terminal: {}
> ```
> 
> ```
> [user@hostname devspace (main)]$ devspace dev
> info Using namespace 'user'
> info Using kube context 'context'
> Ensuring image pull secret for registry: hub.docker.com...
> Couldn't retrieve username for registry  from docker store
> Couldn't retrieve password for registry  from docker store
> local-registry: Starting Local Image Registry
> build_images: build images: error during shouldRebuild check: Dockerfile /Users/user/devspace/Dockerfile missing: stat /Users/user/devspace/Dockerfile: no such file or directory
> fatal exit status 1
> ```
> 
> Based on the docs (https://www.devspace.sh/docs/configuration/images/build#build-configuration) it seems that `dockerfile` is optional, but if a Dockerfile doesn't exist, `devspace dev` fails. I don't see a dockerfile option required in the `devspace dev` doc either (https://www.devspace.sh/docs/configuration/dev/#config-reference). I guess the _option_ is optional, but the _file_ is not optional?

In the [docs](https://www.devspace.sh/docs/configuration/images/#images-dockerfile), It is optional with a default value of `Dockerfile`. If it's inconsistent elsewhere, we should correct that. If you specify any `images` section, DevSpace will build an image.
> 
> If I add a Dockerfile, it will seemingly ignore the ubuntu image, then use whatever is in the Dockerfile. However, it will re-use the last build rather than use a new build if the Dockerfile changes, so subsequent commands rebuild the image, but don't result in a new dev container when the Dockerfile changes.
> 
> To deal with that, I tried explicitly passing the tag in a pipelines section:
> 
> ```
> version: v2beta1
> 
> pipelines:
>   dev: |-
>     run_dependencies --all
>     ensure_pull_secrets --all
>     build_images --all -t $(git describe --always)
>     create_deployments --all
>     start_dev --all
> 
> images:
>   python:
>     image: ubuntu:23.04
> 
> dev:
>   my-dev:
>     imageSelector: ubuntu:23.04
>     terminal: {}
> ```

Using `ubuntu:23.10` in the `images` section of your DevSpace config is an anti-pattern, since the expectation is that you will be able to push the resulting image tag to an image registry. The expected usage is to provide your own image tag:
```yaml
images:
  python:
    image: username/my-python-app
```

And then in your Dockerfile, use the ubuntu as the base image:
```Dockerfile
FROM ubuntu:23.10
```
> 
> And that leads to another hanging situation, even though the secrets it's looking for actually exist:
> 
> ```
> 01:24:46 dev:my-dev Update replaced deployment with patch:
>  {"metadata":{"annotations":{"deployment.kubernetes.io/revision":"2","devspace.sh/container":null}},"spec":{"template":{"metadata":{"annotations":{"devspace.sh/container":null,"devspace.sh/imageSelector":"localhost:31886/ubuntu:9abb179"}}}}}
> 01:24:46 dev:my-dev Waiting for pod to become ready...
> 01:24:46 dev:my-dev Start selecting a single container with selector image selector: localhost:31886/ubuntu:9abb179
> 01:24:57 dev:my-dev Pod app-59fbcd5cff-ljfzr: Back-off restarting failed container (BackOff)
> 01:24:57 dev:my-dev Pod devspace-build-kaniko-kfcjz: Unable to attach or mount volumes: unmounted volumes=[devspace-pull-secrets], unattached volumes=[kube-api-access-2snkh devspace-pull-secrets context]: timed out waiting for the condition (FailedMount)
> 01:24:57 dev:my-dev Pod devspace-build-kaniko-kfcjz: Unable to attach or mount volumes: unmounted volumes=[devspace-pull-secrets], unattached volumes=[devspace-pull-secrets context kube-api-access-2snkh]: timed out waiting for the condition (FailedMount)
> 01:24:57 dev:my-dev Pod devspace-build-kaniko-kfcjz: MountVolume.SetUp failed for volume "devspace-pull-secrets" : secret "devspace-pull-secrets" not found (FailedMount)
> 01:24:57 dev:my-dev DevSpace still couldn't find any Pods that match the selector. DevSpace will continue waiting, but this operation might timeout
> 01:25:07 dev:my-dev DevSpace still couldn't find any Pods that match the selector. DevSpace will continue waiting, but this operation might timeout
> 01:25:18 dev:my-dev DevSpace still couldn't find any Pods that match the selector. DevSpace will continue waiting, but this operation might timeout
> 01:25:28 dev:my-dev DevSpace still couldn't find any Pods that match the selector. DevSpace will continue waiting, but this operation might timeout
> 01:26:10 dev:my-dev DevSpace still couldn't find any Pods that match the selector. DevSpace 
> ```
> 
> Turns out that was due to a pre-existing deployment (devspace operations aren't immutable or idempotent). Did a purge, tried again, complained about no deployment. So I added a deployment section and tried again, and that worked, at first.
> 
> But changing the Dockerfile again still results in it re-using the last cached build. Later it turns out that if I just leave the tag off the `ubuntu:23.04` on all the lines in the config, it works as expected, deploying the currently-built image after the Dockerfile changes. No pipeline or build tag needed, but a deployment section is.

DevSpace is matching on a previously failed deployment because of the `ubuntu:23.10` `imageSelector`, which won't change with the way you've configured it. The expected pattern would be to use your provided image tag `username/my-python-app`, to which DevSpace will append a generated tag that changes whenever the build context or Dockerfile changes.

> 
> As I'm looking for documentation, I'll use your website's search box in the top right hand corner. If I put in "labelSelector", and click the first result, it brings me to the documentation for version 5.x. But I don't realize that until much later, when I realize I can't find this page from the main docs page. The next link after the first one in the search box is for 4.x. No indication from the search box that it's linking to older docs. (incidentally, the 5.x docs are actually easier to read and understand than the 6.x docs)
> 
> labelSelector doesn't even seem to be documented for 6.x: https://www.devspace.sh/docs/configuration/dev/selectors/image#dev-labelSelector
> 
This is good feedback for our docs. A separate issue would be helpful in getting it prioritized. 

> It does seem like there's quite a learning curve to understand how Devspace wants the user to use it, guessing what Devspace doesn't like, trying another thing...

The struggle is real. DevSpace is such a flexible tool that it's getting very difficult to provide documentation for all the different ways it can be used. We have a [loft users slack](https://slack.loft.sh) that could be helpful in getting faster support from the community.
> 
> I really do like this tool, and think it could be a game changer for a lot of teams. But it's taking us weeks/months to work through all the idiosyncrasies to the point where we actually know how to use this tool and can introduce it to our other teams. I've been using it for at least a month straight and I still can't get it to do what I want (just drop me into an Ubuntu container so I can do cloud development work without Docker Desktop installed) without weeks of experimentation and research.
> 
> (for posterity, if anyone is looking for the solution, the following config appears to be required; if any lines are missing or different, you may end up with old cached images, or something else won't work:)
> 
> ```
> version: v2beta1
> 
> images:
>   ubuntu:
>     image: nonsense
>     rebuildStrategy: always
> 
> deployments:
>   app:
>     helm:
>       chart:
>         name: component-chart
>         repo: https://charts.devspace.sh
>       values:
>         containers:
>           - image: nonsense
> 
> dev:
>   my-dev:
>     imageSelector: nonsense
>     terminal: {}
> ```

Here's another way to accomplish your goal of getting an Ubuntu container so I can do cloud development work without Docker Desktop installed:

`devspace.yaml`
```yaml
version: v2beta1

name: ubuntu-no-build

deployments:
  app:
    helm:
      values:
        labels:
          app: my-app
        containers:
          - image: ubuntu:23.10

dev:
  app:
    labelSelector:
      app: my-app
    devImage: ubuntu:23.10
    workingDir: /app
    terminal:
      command: ./devspace_start.sh
    sync:
      - path: ./

```

`devspace_start.sh`
```shell
#!/bin/bash
set +e  # Continue on errors

echo "INSTALL TOOLING"

bash
```

In this case there's no image build needed at all. The downside is that you'll need to wait for the tooling to install each time a terminal is started. Another common pattern is to use the `images` section to build a devImage so that you don't have that additional overhead. Generally you'll also have a production image as well... so combining the two examples would look like:

`devspace.yaml`
```yaml
version: v2beta1

name: ubuntu-no-build

images:
  app-dev:
    image: app-dev
    dockerfile: ./Dockerfile.dev
  app:
    image: app

deployments:
  app:
    helm:
      values:
        containers:
          - image: app

dev:
  app:
    imageSelector: app
    devImage: app-dev
    workingDir: /app
    terminal: {}
    sync:
      - path: ./

```

`Dockerfile`
```Dockerfile
FROM ubuntu:23.10
# production optimized image
```

`Dockerfile.dev`
```Dockerfile
FROM ubuntu:23.10
# development tooling installed
```

When you run `devspace dev` the deployment's image will be swapped out with `app-dev` image automatically. If you were to run `devspace deploy`, the `app` image would get used. We also provide [pre-built images](https://github.com/loft-sh/devspace-containers) to use for `devImage`. If one of those works for your purposes, then you could drop the `app-dev` image.

I encourage you to join the [loft users slack](https://slack.loft.sh), since I think you'll be able to get answers much faster, and also get ideas from others on the different ways to use DevSpace.
